### ISSUE TITLE
Sonoff DIY device does not retain state

### ISSUE THREAD

**Issue by DanBrezeanu** [AlexxIT/SonoffLAN#1121]
I have a Sonoff B05-BL-A60 which has the Sonoff DIY API available. I've added the device just to my network, didn't add it in EWeLink. Tested the device API with Postman and it was working.

I've manually added it in Sonoff LAN with the following configuration:
```yaml
sonoff:
  username: !secret sonoff_username
  password: !secret sonoff_password
  debug: True
  reload: always
  devices:
    1001xxxxxx:
      device_class: light
      name: Sonoff B05
```

The device appeared in HASS as expected, but the light switch does not seem to stay in the "off" position.
Behaviour (switch is in the ON position):
* Clicking the switch moves it to the OFF position and correctly turns off the lights.
* After a couple of seconds, the switch goes back to the ON position  (without clicking on it) and the light is still turned off.

![Imgur](https://imgur.com/9QpE3zO.jpg)

Logs (when clicking the switch from ON to OFF):
```
2023-02-24 10:17:54 [D] 1001xxxxxx => Local4 | {'switch': 'off'} <= {'seq': 2, 'error': 0}
2023-02-24 10:17:59 [D] 1001xxxxxx => Local4 | {'switch': 'off'} <= {'seq': 2, 'error': 0}
2023-02-24 10:18:03 [D] 1001xxxxxx => Local4 | {'switch': 'off'} <= {'seq': 2, 'error': 0}
```

Logs (when clicking the switch when it is in the OFF position \<and the lights actually turn on\> and then back OFF):
```
2023-02-24 10:19:47 [D] 1001xxxxxx => Local4 | {'switch': 'off'} <= {'seq': 2, 'error': 0}
2023-02-24 10:19:48 [D] 1001xxxxxx => Local4 | {'switch': 'on'} <= {'seq': 2, 'error': 0}
2023-02-24 10:19:49 [D] 1001xxxxxx => Local4 | {'switch': 'off'} <= {'seq': 2, 'error': 0}
```

I don't know if this is a configuration issue or an actual bug. I'm running the master@453cbed version of Sonoff LAN.

**Comment by LOVE2CMOL** [comment]
You can try these two PR. 
PR #1134 was submitted to solve the problem of the status not being returned when updated.
According to my preliminary analysis, it is possible that the official server did not receive the ping packets correctly, which led to the failure of pushing the status updates.
So I completed the code according to the documentation provided by the official.

PR #1116 was aimed at solving the disconnection issue.

[#1134](https://github.com/AlexxIT/SonoffLAN/pull/1134)
[#1116](https://github.com/AlexxIT/SonoffLAN/pull/1116)

Currently, I have tested and found that the status can be updated normally for a long time.

**Comment by AlexxIT** [comment]
I don't know if bulbs works fine in DIY mode. Not sure about this.

**Comment by JavonneM** [comment]
Yeah, I'm having the exact same behavior, by default mine detects as `Unknown DIY` and as a switch (I have done no additional configuration).

Some info: 
B05-BL-A60
```
"fwVersion": "1.4.1",
"deviceid": "100XXXXXXX"
```

On startup 
```
2023-06-22 21:09:52.978 DEBUG (MainThread) [custom_components.sonoff] LOCAL mode start
2023-06-22 21:09:52.995 DEBUG (MainThread) [custom_components.sonoff.core.ewelink] 100XXXXXXX UIID 0000 | {'switch': 'on', 'ltype': 'white', 'white': {'br': 100, 'ct': 0}, 'party': {'br': 45, 'r': 254, 'g': 132, 'b': 0, 'tf': 1, 'sp': 1}, 'fwVersion': '1.4.1'}
2023-06-22 21:13:07.639 DEBUG (MainThread) [custom_components.sonoff.core.ewelink.local] 10016c3925 => Local4 | 192.168.0.143:8081 | {'switch': 'off'} <= {'seq': 2, 'error': 0}
```
Not sure if the UIID being `0000` is a problem but will investigate checkout the code base.

Also after playing around a bit I managed to get HA switch stuck in the `off` position and switching to `on` simply flips it back with the `'error': 400` response. It would seem HA is confused by the devices current state and sending a switch state to a state that the device is already in then returns 400 responses.

**Comment by JavonneM** [comment]
Just did some debugging, this seems to be a series of issues that cause the current behavior

1. mDNS service discovery seems to report an the wrong state.

Checking initial state received by mdns/zeroconfig states that the device is initially in the `off` state, my light is in `ON`
```
'txtvers':'1'
'id':'100XXXXXXX'
'type':'diy_light'
'apivers':'1'
'seq':'1'
'data1':'{"switch":"off","ltype":"white","white":{"br":100,"ct":0},"party":{"br":45,"r":254,"g":132,"b":0,"tf":1,"sp":1},"fwVersion":"1.4.1"}'
```
I also copied the zeroconf lib into my own python script, same incorrect result (will test a different lib soon). Querying `http://ip:port/zeroconf/info` gives the correct result
```
{
    "seq": 2,
    "error": 0,
    "data": {
        "switch": "on",
        "ltype": "white",
        "white": {
            "br": 91,
            "ct": 0
        },
        "party": {
            "br": 45,
            "r": 254,
            "g": 132,
            "b": 0,
            "tf": 1,
            "sp": 1
        },
        "fwVersion": "1.4.1",
        "rssi": -73,
        "bssid": "XXXX",
        "deviceid": "XXXX"
    }
}
```
2. There is no mapping for my `B05-BL-A60` and is therefore detected as a default `switch`

As you can see in point 1 mdns device reports as `type: diy_light` however in devices.py https://github.com/AlexxIT/SonoffLAN/blob/8c8ba2e29573331fe14808f6c74f0708fce7ade3/custom_components/sonoff/core/devices.py#L485 its only detects `diylight` and subsequently fails to match in `DIY[ltype]` https://github.com/AlexxIT/SonoffLAN/blob/8c8ba2e29573331fe14808f6c74f0708fce7ade3/custom_components/sonoff/core/devices.py#L494
Understandably, I do not see any type of way to identify these devices from the available apis (ie no uiid present, possibly need to override a device id with a uiid in config to make this easier)

3. After adding a additional entry to the DIY mapping, (`diy_light: uiid 136`) I am able to control the dimmer and color/temp however it seems I am unable to get the light to actually stay in the `on` state.
From what I am seeing
```
2023-06-23 22:43:17 [D] 10016c3925 => Local4 | 192.168.0.143:8081 | {'switch': 'on'} <= {'seq': 2, 'error': 0}
2023-06-23 22:43:21 [D] 10016c3925 => Local4 | 192.168.0.143:8081 | {'switch': 'on'} <= {'seq': 2, 'error': 0}
```
The integration is sending the correct data but for some reason its not updating from what I can see there is no state being saved.  (need to investigate further)

**Comment by JavonneM** [comment]
I've found a way to override `uiid` for a DIY device, added to the `README.md`, this gives my device all the capabilities of  associated as far as I can tell.

https://github.com/AlexxIT/SonoffLAN/pull/1191

**Comment by AlexxIT** [comment]
I think it's better to add `diy_light` to DIY list.

**Comment by JavonneM** [comment]
> I think it's better to add `diy_light` to DIY list.

My only concern is that this is quite a loose definition and I'm not sure if this value varies with different firmware versions or if another device possibly has this same `type` but different functionality ie it doesn't have different colors or something.
I can map this to a basic light that only has on/off functionality, then use the override to be a bit more exact?

**Comment by AlexxIT** [comment]
If there is no such mention in the code, it means that the device with this type has not yet been seen.

**Comment by JavonneM** [comment]
:+1: will add it to the DIY list.

I personally prefer not to use the ewelink app + cloud functionality since I have data concerns and HA has all the control I need on my network.

Are there plans to query and store the device state of DIY devices?

**Comment by AlexxIT** [comment]
No any plans about spending additional time on DIY devices. I think this implementation has bad design.
It's better to use basic ewelink mode or flash devices with ESPHome.

**Comment by isycat** [comment]
> No any plans about spending additional time on DIY devices. I think this implementation has bad design. It's better to use basic ewelink mode or flash devices with ESPHome.

Would you still accept a PR for this?

**Comment by isycat** [comment]
> > I think it's better to add `diy_light` to DIY list.
> 
> My only concern is that this is quite a loose definition and I'm not sure if this value varies with different firmware versions or if another device possibly has this same `type` but different functionality ie it doesn't have different colors or something. I can map this to a basic light that only has on/off functionality, then use the override to be a bit more exact?

The ambiguity would seem to be intentional API design:  a "diy_light" is a "diy_light".
![image](https://github.com/AlexxIT/SonoffLAN/assets/44198750/288eb275-cc21-47a9-af4f-8150785ec11e)
https://sonoff.tech/diy-developer/#7

**Comment by bbs2web** [comment]
I've secretly been hoping that this issue gets fixed, I really appreciate the community and especially Alex's efforts around the Home Assistant integration.

I'm specifically facing this issue with the Sonoff B02-BL-A60 bulbs, they are running firmware 1.6 which uses the LAN control interface.

I can reproduce the issue relatively easily, will collect the information referenced in the integration page and hope that it allows someone with development expertise to fix this. Would definitely buy more bulbs if I could get this problem to be fixed.
